src/OpenTelemetry.Api/Trace/TracerProvider.cs

```diff
@@ -44,5 +46,71 @@ protected TracerProvider()
     /// <param name="version">Version of the instrumentation library.</param>
     /// <returns>Tracer instance.</returns>
     public Tracer GetTracer(string name, string? version = null)
-        => new(new ActivitySource(name ?? string.Empty, version));
+    {
+        var tracers = this.tracers;
+        if (tracers == null)
+        {
+            // Note: Returns a no-op Tracer once dispose has been called.
+            return new(activitySource: null);
+        }
+
+        var key = new TracerKey(name, version);
```

Comment:
I think in this case, it's probably better to use `ConcurrentDictionary` instead of a `HashTable` as it would allow us to avoid this allocation for every lookup. We could use value tuple as the key of the dictionary.

Reply:
I just pushed some updates. You probably aren't going to like them initially but give it a chance! 🤣 

My goal here was to be fast & non-blocking. I wasn't really concerned about the allocation. BUT it is simple enough to eliminate the allocation by reusing an object so that's what I did on the push.

Why not just use `ConcurrentDictionary`? It has slow reads! Considering calls to `GetTracer` will now hit cache 99.999999% of the time I wanted it to be as fast as possible for the read case.

Reply:
> Why not just use ConcurrentDictionary? It has slow reads!

Since this is not the hot-path, I thought it should be okay to use `ConcurrentDictionary` here to favor simplicity over that tiny perf gain coming from `Hashtable`.

Reply:
> I just pushed some updates. You probably aren't going to like them initially but give it a chance! 🤣

I'll check it.

Reply:
> > Why not just use ConcurrentDictionary? It has slow reads!
> 
> Since this is not the hot-path, I thought it should be okay to use `ConcurrentDictionary` here to favor simplicity over that tiny perf gain coming from `Hashtable`.

+1. But the current one is also fine.

Reply:
Updated to use a `ConcurrentDictionary`. Interestingly nothing in `OpenTelemetry.Api` dependency chain is pulling in `System.ValueTuple` for .NET Framework. So to avoid adding a dependency, I used a `readonly record struct` for the key.

Reply:
> I don't think this is accurate! If users are running into this memory leak it is because they have inserted the GetTracer call into their hot path.

I would say that's incorrect usage of the API. For example, for Activity we do not recommend a user to keep recreating the same `ActivitySource` simply to emit an activity on their hot path. If a user is doing that, they should probably fix their usage of the API.

Reply:
I don't disagree with you that it is wrong however we have not provided any documentation and unlike `ActivitySource` `Tracer` does not implement `IDisposable` (which is the standard way in .NET of saying something needs to be correctly managed). If we had `IDisposable` on there I would say it is up to the user to pay attention 😄 Essentially what this PR is doing is compensating for these shortcomings by making it safe regardless. `LoggerFactory` really does the same thing. You can call `CreateLogger` all you want, it will keep you from shooting yourself in the foot by handing back instances from the cache.

Reply:
I'm totally in favor of this PR disposing the underlying `ActivitySource`. My point was that this is not really a "valid" hot path so it's okay to prefer simplicity (ConcurrentDictionary) over small perf gains (Hashtable).

Reply:
I think it actually ended up faster so best of both worlds 😄